docs/design/include/mcmini/MCSharedLibraryWrappers.c

```diff
@@ -0,0 +1,394 @@
+  else {
+	  /*pthread_create_ptr = dlsym(RTLD_NEXT, "pthread_create");
+	  return pthread_create_ptr;*/
+	  int return_value =  __real_pthread_create(pthread, attr, routine, arg);
+	  return return_value;
+	}
+}
+
+int
+pthread_join(pthread_t pthread, void **result)
+{
+  if(mcmini_enabled) return mc_pthread_join(pthread, result);
+  else return __real_pthread_join(pthread, result);
+}
+
+int
+pthread_mutex_init(pthread_mutex_t *mutex,
+                   const pthread_mutexattr_t *mutexattr)
+{  if(mcmini_enabled) return mc_pthread_mutex_init(mutex, mutexattr);
+  else {
+	  //pthread_mutex_init_ptr   = dlsym(RTLD_NEXT, "pthread_mutex_init");
+	  //return pthread_mutex_init_ptr;
+	  int result =  __real_pthread_mutex_init(mutex, mutexattr);
+    return result;
+  }
+}
+
+int pthread_mutex_lock(pthread_mutex_t *mutex)
+{
+  if(mcmini_enabled) return mc_pthread_mutex_lock(mutex);
+  else {  
+	  int return_value = __real_pthread_mutex_lock(mutex);
+	  count_calls++;
+	  /*
+	   * Map the shared memory into the memory with MAP_FIXED_NOREPLACE
+	   * Recording phase.
+	   */
+	  long pagesize = sysconf(_SC_PAGESIZE);
+	  void *desired_address = (void *)((uintptr_t)base_address + 64*count_calls * sizeof(pthread_mutex_t));
+	  //printf("\n -- D.A after count calls %p ---", desired_address);
+	  desired_address = (void *)(((uintptr_t)desired_address / pagesize) * pagesize);
+	  //printf("\n -- D.A after page adjustment %p ---", desired_address);
+	  int retry_count = 0;
+	  size_t offset = sizeof(pthread_mutex_t);
+    while (1) {
+        // Map the shared memory into memory at the desired address
+        mutex = (pthread_mutex_t *)mmap(desired_address, sizeof(pthread_mutex_t), PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED_NOREPLACE, fd, 0);
+        if (mutex == MAP_FAILED) {
+            if (errno == EEXIST) {
+		    retry_count++;
+                // Address already in use, try again with a different address
+                desired_address = (void *)((uintptr_t)desired_address + (retry_count*64)*offset);
+		//printf("\n -- D.A after retry count %p --", desired_address);
+		desired_address = (void *)(((uintptr_t)desired_address / pagesize) * pagesize);
+		//printf("\n ---- new desired address for mmap is %p ------", desired_address); 
+
+      
+                if (retry_count >= 10) {
+                    printf("Maximum retries reached. Unable to find a free address.\n");
+                    exit(EXIT_FAILURE);
+                }
+            } else {
+                perror("mmap");
+                exit(EXIT_FAILURE);
+            }
+        } else {
+            // Mapping successful, break the retry loop
+	    printf("\n Mapped mutex address : %p", mutex);
+            break;
+        }
+    }
+       
+	// append the address to the linked list
+        Rec_list *newNode = (Rec_list *)malloc(sizeof(Rec_list));
+	if (newNode == NULL) {
+    	  perror("malloc");
+    	  exit(EXIT_FAILURE);
+	}
+	newNode->mutex = mutex;
```

Comment:
Where exactly do you track that `mutex` is now locked, i.e. changed state as a result of this call?